File: ChatService/src/main/scala/APIs/UserService/RegisterUserMessage.scala
package APIs.UserService

import Common.API.API
import Global.ServiceCenter.UserServiceCode

import io.circe.{Decoder, Encoder, Json}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import io.circe.syntax.*
import io.circe.parser.*
import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

import com.fasterxml.jackson.core.`type`.TypeReference
import Common.Serialize.JacksonSerializeUtils

import scala.util.Try

import org.joda.time.DateTime
import java.util.UUID
import Objects.UserService.UserRole

/**
 * RegisterUserMessage
 * desc: 用户注册，生成唯一的userID，创建用户记录并返回userID。
 * @param username: String (用户名，用于标识和登录用户。)
 * @param password: String (密码，用于用户登录和验证。)
 * @param role: UserRole:1132 (用户角色，例如普通用户或商家。)
 * @return userID: String (新创建用户的唯一标识。)
 */

case class RegisterUserMessage(
  username: String,
  password: String,
  role: UserRole
) extends API[String](UserServiceCode)



case object RegisterUserMessage{
    
  import Common.Serialize.CustomColumnTypes.{decodeDateTime,encodeDateTime}

  // Circe 默认的 Encoder 和 Decoder
  private val circeEncoder: Encoder[RegisterUserMessage] = deriveEncoder
  private val circeDecoder: Decoder[RegisterUserMessage] = deriveDecoder

  // Jackson 对应的 Encoder 和 Decoder
  private val jacksonEncoder: Encoder[RegisterUserMessage] = Encoder.instance { currentObj =>
    Json.fromString(JacksonSerializeUtils.serialize(currentObj))
  }

  private val jacksonDecoder: Decoder[RegisterUserMessage] = Decoder.instance { cursor =>
    try { Right(JacksonSerializeUtils.deserialize(cursor.value.noSpaces, new TypeReference[RegisterUserMessage]() {})) } 
    catch { case e: Throwable => Left(io.circe.DecodingFailure(e.getMessage, cursor.history)) }
  }
  
  // Circe + Jackson 兜底的 Encoder
  given registerUserMessageEncoder: Encoder[RegisterUserMessage] = Encoder.instance { config =>
    Try(circeEncoder(config)).getOrElse(jacksonEncoder(config))
  }

  // Circe + Jackson 兜底的 Decoder
  given registerUserMessageDecoder: Decoder[RegisterUserMessage] = Decoder.instance { cursor =>
    circeDecoder.tryDecode(cursor).orElse(jacksonDecoder.tryDecode(cursor))
  }


}

File: Plugins/UserService/APIs/RegisterUserMessage.ts
/**
 * RegisterUserMessage
 * desc: 用户注册，生成唯一的userID，创建用户记录并返回userID。
 * @param username: String (用户名，用于标识和登录用户。)
 * @param password: String (密码，用于用户登录和验证。)
 * @param role: UserRole:1132 (用户角色，例如普通用户或商家。)
 * @return userID: String (新创建用户的唯一标识。)
 */
import { TongWenMessage } from 'Plugins/TongWenAPI/TongWenMessage'
import { UserRole } from 'Plugins/UserService/Objects/UserRole';


export class RegisterUserMessage extends TongWenMessage {
    constructor(
        public  username: string,
        public  password: string,
        public  role: UserRole
    ) {
        super()
    }
    getAddress(): string {
        return "127.0.0.1:10013"
    }
}

